compiler/parser/RdrHsSyn.hs

```diff
@@ -2533,32 +2533,35 @@ data TemplateBodyDecl
   | KeyDecl (Located KeyData)
   | MaintainerDecl (LHsExpr GhcPs)
 
+data TemplateBodyDecls =
+  TemplateBodyDecls {
+      tbdEnsures :: [LHsExpr GhcPs]
+    , tbdSignatories :: [LHsExpr GhcPs] -- (list of lists)
+    , tbdObservers :: [LHsExpr GhcPs] -- (list of lists)
+    , tbdAgreements :: [LHsExpr GhcPs]
+    , tbdControlledChoiceGroups :: [Located (LHsExpr GhcPs, Located [Located ChoiceData])]
+    , tbdLetBindings :: [LHsLocalBinds GhcPs]
+    , tbdFlexChoices :: [Located FlexChoiceData]
+    , tbdKeys :: [Located KeyData]
+    , tbdMaintainers :: [LHsExpr GhcPs] -- (list of lists)
+    }
+
 -- | Classify a list of template body declarations.
-extractTemplateBodyDecls ::
-     [Located TemplateBodyDecl]
-  -> ( [LHsExpr GhcPs] -- ensure
-     , [LHsExpr GhcPs] -- signatories (list of lists)
-     , [LHsExpr GhcPs] -- observers (list of lists)
-     , [LHsExpr GhcPs] -- agreement
-     , [Located (LHsExpr GhcPs, Located [Located ChoiceData])] -- controlled choice groups
-     , [LHsLocalBinds GhcPs] -- let bindings
-     , [Located FlexChoiceData] -- flexible choices
-     , [Located KeyData] -- key
-     , [LHsExpr GhcPs] -- maintainers (list of lists)
-     )
-extractTemplateBodyDecls = foldl extract ([], [], [], [], [], [], [], [], [])
+extractTemplateBodyDecls :: [Located TemplateBodyDecl] -> TemplateBodyDecls
```

Comment:
Why not make this type a monoid? Then it's nice and simples. You might even be able to derive Monoid

Reply:
Ooh. Thank-you for this tip! I've rolled it out by hand because I don't think I can derive without adding a package dependency.

Reply:
@neil-da Can I impose on you for another look over this please (the `Monoid` bits - is there any more brevity I can obtain)?